File: AutoSync/42-trapping-rain-water/solution.py
class Solution:
    def trap(self, height):
        ans = 0;
        left = 0;
        right = len(height) - 1;
        leftMax = height[0];
        rightMax = height[-1];

        while (left < right):
            leftMax = max(height[left], leftMax);
            rightMax = max(height[right], rightMax);
            if leftMax < rightMax:
                ans += leftMax - height[left];
                left += 1;
            else:
                ans += rightMax - height[right];
                right -= 1;
        return ans;

File: AutoSync/42-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        return Method(height);
    }
    public static int Method(int[] heights){
        Deque<Integer> deque = new ArrayDeque<>();
        int i = 0;
        int peek;
        int h ,w, ans=0;
          while(i < heights.length){
            if (deque.isEmpty() || heights[deque.peek()] > heights[i]){
                deque.push(i);
            }else {
                peek = deque.pop();
                if (deque.isEmpty()){
                    continue;
                }
                h = Math.min(heights[deque.peek()],heights[i])-heights[peek];
                w = (i-deque.peek()-1);
                ans+=h*w;
                i--;
            }
            i++;
        }
        return ans;
    }
}
